### PR TITLE
vm-manage-gpu: Update mdevs

### PR DESCRIPTION
**Pull request by ljelinkova**
The backend side of the mdev (vGPUs) has been changed from using custom properties to vm devices + a new driver parameters field has been added to the dialog. In addition to that, the dialog has been moved from host devices tab to vm devices tab.

image

When there are inconsistencies in the data in the database (e.g. the user inserted the devices using REST API), the following warnings are displayed.
image

Bug-Url: https://bugzilla.redhat.com/show_bug.cgi?id=1987121

### CODE REVIEW

File: src/integrations/buttons.js
Comment (mz-pdm) [inline src/integrations/buttons.js:17]: I cannot see the button there (nor in the host devices tab).  Perhaps a problem on my side?

File: src/modals/vm-manage-gpu/GpuDataProvider.js
Comment (mz-pdm) [inline src/modals/vm-manage-gpu/GpuDataProvider.js:81]: To be consistent with backend processing, it would be better to check if any of the devices has nodisplay, rather than the first one.

File: src/modals/vm-manage-gpu/GpuDataProvider.js
Comment (mz-pdm) [inline src/modals/vm-manage-gpu/GpuDataProvider.js:98]: And here it may be better to take the first non-empty value (if there is any).

File: src/modals/vm-manage-gpu/ManageGpuModal.js
Comment (mz-pdm) [inline src/modals/vm-manage-gpu/ManageGpuModal.js:81]: Will it silently delete driverParams if present and version is < 4.7?  This can happen e.g. when a VM has custom compatibility level set and later unset, without noticing.

If this is the case then it would be better to keep the parameters and perhaps indicate some hint that they may not be applied.  Note that driver parameters are sent to the host in the backend even when version is 4.7 and they may be applied or ignored there, depending on the Vdsm version running on the host.  So it is OK to not allow defining the parameters on < 4.7 but it may not be very user friendly just to drop them silently if they are already defined.

File: src/intl/messages.js
Comment (mz-pdm) [inline src/intl/messages.js:871]: The driver parameters are allowed to be set only for NVidia GPUs. I'm not sure whether we support vGPUs for other vendors but in case we do, the limitation should be mentioned here.

File: src/integrations/buttons.js
Comment (ahadas) [reply src/integrations/buttons.js:17]: @ljelinkova  would this be solved by https://github.com/oVirt/ovirt-engine/pull/173?

File: src/intl/messages.js
Comment (ahadas) [inline src/intl/messages.js:791]: Missing*

File: src/intl/messages.js
Comment (ahadas) [inline src/intl/messages.js:797]: Missing*

File: src/intl/messages.js
Comment (ahadas) [reply src/intl/messages.js:871]: The only place I found in the documentation that is related to vGPU is titled `SETTING UP AN NVIDIA GPU FOR A VIRTUAL MACHINE IN RED HAT VIRTUALIZATION` so I think that's fine

File: src/modals/vm-manage-gpu/ManageGpuModal.js
Comment (ahadas) [reply src/modals/vm-manage-gpu/ManageGpuModal.js:81]: > Note that driver parameters are sent to the host in the backend even when version is 4.7 and they may be applied or ignored there, depending on the Vdsm version running on the host. So it is OK to not allow defining the parameters on < 4.7 but it may not be very user friendly just to drop them silently if they are already defined.

If I understand this part correctly then that's actually an argument for clearing it - so it won't be sent when the compatibility level is changed back to < 4.7. but that's really not that interesting case, let's not hold this PR on that

File: src/integrations/buttons.js
Comment (ljelinkova) [reply src/integrations/buttons.js:17]: Yes

File: src/modals/vm-manage-gpu/ManageGpuModal.js
Comment (ljelinkova) [reply src/modals/vm-manage-gpu/ManageGpuModal.js:81]: I think we should not handle the driver params for VMs < 4.7 at all. I've changed the patch so that the values are not displayed nor stored if the VM has CL < 4.7.

**Comment by sjd78** [comment]
@ljelinkova - What BZ is this PR for?  Please add a Bug-Url link in the description.

**Review by rszwajko — COMMENTED** [review]
Looks good!

File: src/modals/vm-manage-gpu/GpuDataProvider.js
Comment (sjd78) [reply src/modals/vm-manage-gpu/GpuDataProvider.js:230]: I'm less concerned about optimizing fetches here than making sure they are ok to read and each one has things in the right order.  If the calls are independent, for sure go parallel.  But they seem pretty dependent so keeping the fetches serial is ok with me.

**Review by sjd78 — CHANGES_REQUESTED** [review]
Everything look pretty good.  @rszwajko has some good code clarity suggestion.  Only a few comments, but they're more code clarity than anything.

Two non-code things:
1. Can you list the BZ in the description please?
2. Please squash your commits.  We will "Squash and Merge" ultimately, but it is nice to have a single commit message that isn't modified by later comments.

@sgratch, @rszwajko - I'm going to submit my review as "request changes".  If time becomes a factor, I'm ok with the PR being merged without the code clarity comments being addressed.  In that case, just dismiss my review and move it forward.  A quick followup PR to address them would be reasonable.

**Comment by ahadas** [comment]
> 1. Can you list the BZ in the description please?

Added

**Review by ahadas — APPROVED** [review]

**Review by sjd78 — APPROVED** [review]
With addressing the code clarity comments of the current reviews in a future PR, and pending successful OST, this PR LGTM and is ready to merge.

File: src/utils/CompatibilityVersion.js
Comment (sjd78) [reply src/utils/CompatibilityVersion.js:7]: (future PR)

File: src/modals/vm-manage-gpu/ManageGpuModal.js
Comment (sjd78) [reply src/modals/vm-manage-gpu/ManageGpuModal.js:72]: (future PR)

File: src/modals/vm-manage-gpu/GpuDataProvider.js
Comment (sjd78) [reply src/modals/vm-manage-gpu/GpuDataProvider.js:303]: (future PR)

File: src/modals/vm-manage-gpu/GpuDataProvider.js
Comment (sjd78) [reply src/modals/vm-manage-gpu/GpuDataProvider.js:230]: (future PR)

File: src/modals/vm-manage-gpu/GpuDataProvider.js
Comment (sjd78) [reply src/modals/vm-manage-gpu/GpuDataProvider.js:145]: (future PR)

File: src/modals/vm-manage-gpu/GpuDataProvider.js
Comment (sjd78) [reply src/modals/vm-manage-gpu/GpuDataProvider.js:104]: (future PR)

File: src/modals/vm-manage-gpu/GpuDataProvider.js
Comment (sjd78) [reply src/modals/vm-manage-gpu/GpuDataProvider.js:81]: (future PR)

File: src/modals/vm-manage-gpu/GpuDataProvider.js
Comment (sjd78) [reply src/modals/vm-manage-gpu/GpuDataProvider.js:68]: (future PR)

**Pull request merged**
